File: Sources/AdbShell/Services/MacroRunner.swift
import Foundation

struct MacroRunResult: Identifiable {
    let id = UUID()
    let argsLine: String
    let output: String
    let isError: Bool
}

/// Выполняет шаги макроса и резолвит переменные вида `${NAME}` в их строках
/// аргументов — общий движок, используется и из UI (MacroView), и из
/// автозапуска по подключению устройства (DevicesViewModel), поэтому логика
/// живёт в одном месте, а не только внутри View.
enum MacroRunner {
    private static let variablePattern = try? NSRegularExpression(pattern: "\\$\\{([A-Za-z0-9_]+)\\}")

    /// Имена переменных, встречающихся в шагах макроса, в порядке первого появления, без повторов.
    static func variableNames(in macro: Macro) -> [String] {
        var seen: [String] = []
        for step in macro.steps {
            for name in matches(in: step.argsLine) where !seen.contains(name) {
                seen.append(name)
            }
        }
        return seen
    }

    private static func matches(in line: String) -> [String] {
        guard let variablePattern else { return [] }
        let range = NSRange(line.startIndex..., in: line)
        return variablePattern.matches(in: line, range: range).compactMap { match in
            guard match.numberOfRanges > 1, let r = Range(match.range(at: 1), in: line) else { return nil }
            return String(line[r])
        }
    }

    /// Подставляет значения переменных в строку аргументов. Переменная без
    /// значения в словаре остаётся как есть (`${NAME}`) — так ошибка видна
    /// в выводе команды, а не проглатывается молча.
    static func resolve(_ argsLine: String, variables: [String: String]) -> String {
        var result = argsLine
        for (name, value) in variables {
            result = result.replacingOccurrences(of: "${\(name)}", with: value)
        }
        return result
    }

    /// Выполняет шаги по очереди, вызывая `onStep` после каждого. Если у макроса
    /// включён `abortOnFirstFailure`, останавливается на первом же шаге с ошибкой.
    @discardableResult
    static func run(
        _ macro: Macro,
        serial: String,
        service: ADBService,
        variables: [String: String],
        onStep: @escaping (MacroRunResult) -> Void
    ) async -> Bool {
        for step in macro.steps {
            let resolvedLine = resolve(step.argsLine, variables: variables)
            let tokens = resolvedLine.split(separator: " ").map(String.init)
            guard !tokens.isEmpty else { continue }
            do {
                let result = try await service.run(tokens, serial: serial)
                onStep(MacroRunResult(argsLine: resolvedLine, output: result.combined, isError: result.exitCode != 0))
                if result.exitCode != 0, macro.abortOnFirstFailure { return false }
            } catch {
                onStep(MacroRunResult(argsLine: resolvedLine, output: error.localizedDescription, isError: true))
                if macro.abortOnFirstFailure { return false }
            }
        }
        return true
    }
}

File: desktop/src/main/macros/MacroRunner.ts
// Порт MacroRunner.run(...) из Sources/AdbShell/Services/MacroRunner.swift —
// выполняет шаги макроса по очереди, вызывая onStep после каждого. Общий
// движок: используется и запуском из вкладки "Макросы" (main.ts,
// macros:run), и автозапуском по подключению устройства
// (renderer.ts triggerAutorunMacros -> macros:run с variables: {}).
//
// В отличие от Swift-версии, результаты шагов здесь не транслируются в
// renderer по мере выполнения (не через push-события, как логкат) — весь
// список результатов возвращается одним ответом, когда макрос закончил
// работу целиком. Для типичных макросов (несколько adb-команд) это не
// заметно; сознательное упрощение, чтобы не городить run-id/событийную
// корреляцию ради этой волны.

import { AdbService } from '../adb/AdbService';
import { combinedOutput } from '../adb/types/ProcessResult';
import { Macro, MacroRunResult } from '../adb/types/Macro';
import { resolveVariables } from './macroRunnerLogic';

export interface MacroRunOutcome {
  completedFully: boolean;
  results: MacroRunResult[];
}

export async function runMacro(
  macro: Macro,
  serial: string,
  service: AdbService,
  variables: Record<string, string>
): Promise<MacroRunOutcome> {
  const results: MacroRunResult[] = [];
  for (const step of macro.steps) {
    const resolvedLine = resolveVariables(step.argsLine, variables);
    const tokens = resolvedLine.split(' ').filter((t) => t.length > 0);
    if (tokens.length === 0) continue;
    try {
      const result = await service.run(tokens, { serial });
      const isError = result.exitCode !== 0;
      results.push({ argsLine: resolvedLine, output: combinedOutput(result), isError });
      if (isError && macro.abortOnFirstFailure) return { completedFully: false, results };
    } catch (error) {
      results.push({ argsLine: resolvedLine, output: (error as Error).message, isError: true });
      if (macro.abortOnFirstFailure) return { completedFully: false, results };
    }
  }
  return { completedFully: true, results };
}
